server/src/main/java/com/hedera/block/server/persistence/cache/BNLinkedHashMap.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Hedera Block Node
+ *
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.block.server.persistence.cache;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * A custom implementation of a LinkedHashMap that removes the eldest entry when the size exceeds a
```

Comment:
We should probably try to avoid writing our own cache layer.
There are a lot of excellent cache implementations, some with write-through options, and it would be better, in my opinion, for us to use one of those instead.

Reply:
Do you mean a third-party cache library?  There's been a strong mandate to reduce third-party dependencies to minimize security concerns.

Reply:
I understand that, but at the same time, writing a cache layer is enough of a task (if it's to work well) that I think we should at least consider using a third-party library for this case. We simply cannot write absolutely everything in the first release, and perhaps we start with third-party cache then write our own in a future release.

Reply:
@rbair23 What are your thoughts on leveraging a third-party library for caching here?

Reply:
Actually, now that I read through this more, I wonder if we need a cache of any kind to begin with.
Storing whole blocks in memory isn't great (a block is far too big), and I don't see much value to storing a partial set of individual block items in memory in the order they arrive.  If we just stream direct to persistence and only retain the calculated _hash_ value, we can assure persistence and verify a block proof without the memory requirement exceeding a few block _items_.
If we have stream subscribers we should be running a _persistence follower_ which may send immediately, or may wait for block proof before sending out the items for the next block (depending on what was requested).

Reply:
Block items and block proofs aren't part of this PR.  This solution was originally from the first demo where the producers wrote to the cache and the file system and the consumers read the live and historic blocks from a single interface (pulled from the file system as necessary).  Given the refactoring effort discussed in stand-up today, we will have N consumers reading the latest block from a data structure (perhaps this cache) rather than fetching from the file system.  I'll leave this solution in place for now but we will undoubtedly revisit it.